### PR TITLE
`Integrated code lifecycle`: Persist finished build jobs

### PR DESCRIPTION
**Pull request by laurenzfb**
#### General

- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

We want to store finished build jobs in a database table for several purposes, e.g., displaying latest build job in the client, statistics. 

### Description

This PR adds the functionality to persist build jobs, once they are finished (succeeded, cancelled or failed) in the database.

### Steps for Testing

:rotating_light: :rotating_light: **Only test locally, contains database changes** :rotating_light: :rotating_light:
**Undo database changes**: `drop table build_job;`

Prerequisites:
- Programming exercise

1. Create a few submissions
2. Once they are processed, check `build_job` in your database for entries and verify the values

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Server
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [BuildJob.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5932/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain/BuildJob.html) | 100% | :white_check_mark: |
| [VcsRepositoryUri.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5932/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain/VcsRepositoryUri.html) | 100% | :white_check_mark: |
| [BuildJobResult.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5932/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.enumeration/BuildJobResult.html) | 100% | :white_check_mark: |
| [RepositoryType.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5932/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.enumeration/RepositoryType.html) | 100% | :white_check_mark: |
| [ContinuousIntegrationTriggerService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5932/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.ci/ContinuousIntegrationTriggerService.html) | 100% | :white_check_mark: |
| [LocalCIBuildJobExecutionService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5932/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.localci/LocalCIBuildJobExecutionService.html) | 69% | Will be improved in https://github.com/ls1intum/Artemis/pull/7861 and #7871 :white_check_mark: |
| [LocalCIBuildJobManagementService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5932/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.localci/LocalCIBuildJobManagementService.html) | 70% | '' :white_check_mark: |
| [LocalCISharedBuildJobQueueService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5932/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.localci/LocalCISharedBuildJobQueueService.html) | 72% | '' :white_check_mark: |
| [LocalCITriggerService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5932/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.localci/LocalCITriggerService.html) | 100% | :white_check_mark: |
| [LocalCIBuildJobQueueItem.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5932/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.localci.dto/LocalCIBuildJobQueueItem.html) | 100% | :white_check_mark: |
| [LocalVCServletService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5932/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.connectors.localvc/LocalVCServletService.html) | 90% | :white_check_mark: |
| [JavaTemplateUpgradeService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5932/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.programming/JavaTemplateUpgradeService.html) | 97% | :white_check_mark: |
| [ProgrammingSubmissionService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5932/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.programming/ProgrammingSubmissionService.html) | 89% | :white_check_mark: |
| [ProgrammingTriggerService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5932/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.programming/ProgrammingTriggerService.html) | 86% | :white_check_mark: |

## Summary by CodeRabbit

- **New Features**
	- Introduced the `BuildJob` entity to represent build jobs with comprehensive details.
	- Expanded visibility of the `repositorySlug` method for external access.
	- Added `SUCCESSFUL`, `FAILED`, `ERROR`, and `CANCELED` as build job result options.
	- Included `USER` as a new repository type.

- **Enhancements**
	- Improved `ContinuousIntegrationTriggerService` to handle different repository types.
	- Updated `LocalCIBuildJobExecutionService` and `LocalCIBuildJobManagementService` to support new build job parameters and logic.
	- Enhanced `LocalCISharedBuildJobQueueService` for better build job result handling and database interactions.
	- Refined `LocalCITriggerService` to use `RepositoryType` enum and handle null values more effectively.
	- Adjusted `ProgrammingSubmissionService` and `ProgrammingTriggerService` to utilize new build trigger parameters.

- **Refactor**
	- Renamed parameters and updated logic to handle auxiliary repositories in `LocalVCServletService`.
	- Modified `JavaTemplateUpgradeService` to account for new repository types in template upgrades.
	- Overloaded methods in `ProgrammingTriggerService` for improved flexibility with repository types.

- **Tests**
	- Extended test suites to cover new `BuildJob` persistence and repository type checks.

### CODE REVIEW

**Review by reschandreas — DISMISSED** [review]
tested locally works as expected, code also lgmt
Screenshot 2024-01-14 at 15 49 08

**Review by milljoniaer — DISMISSED** [review]
demonstrated in testing session 🚀

**Review by rstief — DISMISSED** [review]
Demonstrated in testing session, looked good.

**Review by mateusmm01 — DISMISSED** [review]
Code LGTM !

File: src/main/resources/config/liquibase/changelog/20240112190652_changelog.xml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        <createTable tableName="build_job">
+            <column name="id" type="bigint" autoIncrement="true">
+                <constraints primaryKey="true" nullable="false"/>
+            </column>
+            <column name="name" type="varchar(255)"/>
+            <column name="exercise_id" type="bigint"/>
+            <column name="course_id" type="bigint"/>
+            <column name="participation_id" type="bigint"/>
+            <column name="build_agent_address" type="varchar(255)"/>
+            <column name="build_start_date" type="timestamp"/>
```
Comment (krusche) [inline src/main/resources/config/liquibase/changelog/20240112190652_changelog.xml:17]: datetime(3)

File: src/main/resources/config/liquibase/changelog/20240112190652_changelog.xml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+            <column name="build_completion_date" type="timestamp"/>
```
Comment (krusche) [inline src/main/resources/config/liquibase/changelog/20240112190652_changelog.xml:18]: datetime(3)

File: src/main/resources/config/liquibase/changelog/20240112190652_changelog.xml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+            <column name="repository_type" type="varchar(10)"/>
+            <column name="repository_name" type="varchar(255)"/>
+            <column name="commit_hash" type="varchar(40)"/>
+            <column name="retry_count" type="int"/>
```
Comment (krusche) [inline src/main/resources/config/liquibase/changelog/20240112190652_changelog.xml:22]: tiny_int

File: src/main/resources/config/liquibase/changelog/20240112190652_changelog.xml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+            <column name="priority" type="int"/>
```
Comment (krusche) [inline src/main/resources/config/liquibase/changelog/20240112190652_changelog.xml:23]: tiny_int

File: src/main/java/de/tum/in/www1/artemis/repository/BuildJobRepository.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+    List<BuildJob> findAllByExerciseId(Long exerciseId);
+
+    List<BuildJob> findAllByParticipationId(Long participationId);
+
+    Optional<BuildJob> findFirstByParticipationIdOrderByBuildStartDateDesc(Long participationId);
+
+    List<BuildJob> findAllByBuildAgentAddress(String buildAgentAddress);
+
+    @Query("SELECT buildjob FROM BuildJob buildjob WHERE buildjob.dockerImage = :dockerImage ORDER BY buildjob.buildStartDate DESC")
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/repository/BuildJobRepository.java:23]: please use the syntax with multi line strings and make sure to use the correct indentation

**Review by krusche — CHANGES_REQUESTED** [review]
Please add indices to the DB, because otherwise the search will be super slow later if there are several million entries in there.
So everything you could search for in BuildJobRepository needs an index, especially courseId, exerciseId, participationId, buildAgentAddress and the two dates

Please also adjust the data types (see comments in Github) and please change the query to multi line string with correct indentation

**Review by mateusmm01 — APPROVED** [review]
Code changes lgtm

**Review by rstief — APPROVED** [review]
reapprove after changes

**Review by reschandreas — APPROVED** [review]
reapprove after code changes

**Review by milljoniaer — APPROVED** [review]
reapprove after code changes

**Review by krusche — APPROVED** [review]

**Pull request merged**